Data/IntMap/Internal.hs

```diff
@@ -937,6 +944,48 @@ updateLookupWithKey _ _ Nil = (Nothing,Nil)
 -- | /O(min(n,W))/. The expression (@'alter' f k map@) alters the value @x@ at @k@, or absence thereof.
 -- 'alter' can be used to insert, delete, or update a value in an 'IntMap'.
 -- In short : @'lookup' k ('alter' f k m) = f ('lookup' k m)@.
+
+#if __GLASGOW_HASKELL__ >= 802
+alter :: (Maybe a -> Maybe a) -> Key -> IntMap a -> IntMap a
+alter f !k t = case alter# f k t of
+  (# (# #) | #) -> t
+  (# | t' #) -> t'
+{-# INLINE alter #-}
+
+
+-- Internal implementation which keeps track of whether or not the intmap was
+-- modified using an unboxed sum (Maybe).
+--
+-- If no modifications are made to the map (# (# #) | #) is returned, otherwise
+-- (# | newMap #) is returned.
+alter# :: (Maybe a -> Maybe a) -> Key -> IntMap a -> (# (# #) | IntMap a #)
+alter# f !k t@(Bin p m l r)
+  | nomatch k p m = case f Nothing of
+                      Nothing -> (# (# #) | #)
+                      Just x -> (# | link k (Tip k x) p t #)
```

Comment:
I don't think you're being strict enough with your return values.

Reply:
Do you mean something like the following?

```haskell
let !t' = link k (Tip k x) p t in (# | t' #)
```

Or something else?

Reply:
Yes, but see the link below for a clearer way.